src/tuwien/auto/calimero/knxnetip/util/Srp.java

```diff
@@ -0,0 +1,288 @@
+			this.type = type;
+		}
+	}
+
+	private final int size;
+	private final boolean mandatory;
+	private final Type type;
+	private final byte[] data;
+
+	private static final int SrpHeaderSize = 2;
+
+	/**
+	 * Creates a new SRP and initializes basic fields.
+	 * <p>
+	 *
+	 * @param srpType one of the search request parameter block types {@see Type}
+	 * @param isMandatory to be evaluated by a KNXnet/IP router or server device
+	 * @param data byte array containing additional data, ignored for select by programming mode SRPs
+	 */
+	public Srp(final Srp.Type srpType, final boolean isMandatory, final byte... data) {
+		type = srpType;
+		mandatory = isMandatory;
+
+		switch (srpType) {
+		case InvalidSrp:
+		case SelectByProgrammingMode:
+			size = SrpHeaderSize;
+			this.data = new byte[0];
+			break;
+		case SelectByMacAddress:
+			if (data.length != 6)
+				throw new KNXIllegalArgumentException("MAC size does not match expected value");
+			size = SrpHeaderSize + 6;
+			this.data = data;
+			break;
+		case SelectByService:
+			if (data.length != 2)
+				throw new KNXIllegalArgumentException("service info size does not match expected value");
+			size = SrpHeaderSize + 2;
+			this.data = data;
+			break;
+		case RequestDibs:
+			if (data.length < 1)
+				throw new KNXIllegalArgumentException("requested DIBs size does not match expected value");
+			this.data = new byte[(data.length % 2) == 0 ? data.length : data.length + 1];
+			System.arraycopy(data, 0, this.data, 0, data.length);
+			size = SrpHeaderSize + this.data.length;
+			break;
+		default:
+			throw new KNXIllegalArgumentException("illegal SRP type + 0x" + Integer.toHexString(srpType.getValue()));
+		}
+	}
+
+	/**
+	 * Creates a new SRP out of a byte array.
+	 * <p>
+	 *
+	 * @param data byte array containing SRP structure
+	 * @param offset start offset of SRP in <code>data</code>
+	 * @throws KNXFormatException if no SRP found or invalid structure
+	 */
+	public Srp(final byte[] data, final int offset) throws KNXFormatException {
+		if (data.length - offset < SrpHeaderSize)
+			throw new KNXFormatException("buffer too short for SRP header");
+
+		size = data[offset] & 0xff;
+		if (size > data.length - offset)
+			throw new KNXFormatException("SRP size bigger than actual data length", size);
+
+		mandatory = (data[offset + 1] & 0x80) == 0x80;
+		type = Srp.Type.from(data[offset + 1] & 0x7f);
+
+		if ((size - SrpHeaderSize) > 0) {
+			this.data = new byte[size - SrpHeaderSize];
+			System.arraycopy(data, offset + SrpHeaderSize, this.data, 0, size - SrpHeaderSize);
+		} else {
+			this.data = new byte[0];
+		}
+	}
+
+	/**
+	 * Returns a search request parameter block to limit the extended search request to KNXnet/IP router or
+	 * server devices where programming mode is currently enabled. The mandatory flag of the SRP is not set.
+	 *
+	 * @return search request parameter block for devices currently in programming mode
+	 */
+	public static Srp inProgrammingMode() {
```

Comment:
I know I wrote the `inProgrammingMode` initially, now that I see all your other ones are `with...` -> if you like to keep the `with...` form for consistency that's fine: `withProgrammingMode`. Decide as you like it better ;)

Reply:
Fixed.